### PR TITLE
Associate user with session

### PR DESCRIPTION
**Pull request by ThatJuanGuy**
## Description

This adds the functionality for user's information to be saved between sessions and computers if they are logged in. Before, all their session data would be destroyed upon logout. Now, it isn't. This is done by saving it in a different session that is never destroyed (the data session) which can then be accessed via `retrieve_data_session` when necessary. If a user is not logged in, `retrieve_data_session` will return the usual `request.session`.

## How to test

I wrote some tests for `retireve_data_session` so running those and checking if they pass would be a good place to start. Make sure to migrate the database because this PR has a new model that links a user to their data session. Aside from the manual tests, I would just try calling all the APIs we have and making sure they don't return forbidden or not work or something. I would also save a schedule, logout, and then log back in and see if it's still there since that was the main point of adding this in the first place.

## Related tasks

Closes # I don't think this had an issue actually. Should I make one? Maybe #267 but that's kind of a stretch and I think Ryan already did that one.

### CODE REVIEW

**Comment by gannonprudhomme** [comment]
Can you update the branches for this (and base this off of the respective branch, presumably `logout-button`) so we don't see the commits that are also in #338?

**Comment by ThatJuanGuy** [comment]
Yes. How do I do that? Do I rebase?

**Comment by gannonprudhomme** [comment]
yeah basically you should update the base branch (`logout-button`) by either merging or rebasing with `master`, then update this branch with `logout-button`. At that point you should change the base branch for this PR from `master` to `logout-button`, so it doesn't show the commits that are in `logout-button` but are not in `master`

**Comment by ThatJuanGuy** [comment]
Ok. So I started rebasing but got really confused and aborted the rebase. What does it mean by current and incoming changes? Since I'm pretty rebasing works by applying all the commits sequentially on top of the latest version of another branch, is current the latest version of the branch I want to rebase onto master and incoming an older commit from the same branch?

**Comment by gannonprudhomme** [comment]
> Ok. So I started rebasing but got really confused and aborted the rebase. What does it mean by current and incoming changes? Since I'm pretty rebasing works by applying all the commits sequentially on top of the latest version of another branch, is current the latest version of the branch I want to rebase onto master and incoming an older commit from the same branch?

Yeah I feel you honestly current and incoming changes always confuse me and I typically have to figure it out while I'm working on it. A normal merge might be easier in this case though

To answer your question, I believe current changes is the older one, and incoming change is the commit to be applied. So for this situation, if you rebase `logout-button` onto `master`, the first conflict's current change will be the one from `master`, and incoming will be from `logout-button`, since the incoming changes are the changes you're (re)applying from the original branch (`logout-button`)

**Comment by ThatJuanGuy** [comment]
It has been done.

**Comment by ThatJuanGuy** [comment]
Currently, only the first course added is saved (or at least appears) after saving courses.

File: autoscheduler/user_sessions/utils/retrieve_data_session.py
Comment (ryan-conn) [inline autoscheduler/user_sessions/utils/retrieve_data_session.py:43]: Not saying that you should change your context manager to be exactly how I did it here, but since I made one for another project really recently I thought it might be helpful to show what I did just for comparison:
```python
@contextmanager
def session_for_request(request):
    """ Gets the permanent session for the user that made a request """
    user_id = request.user.id
    try:
        # Not logged in, use normal session
        if user_id is None:
            session = request.session
            yield session

        # Logged in, attempt to use existing session
        else:
            session_key = AuthUserSession.objects.select_related('session').get(
                auth_user_id=user_id
            ).session.session_key
            session = SessionStore(session_key=session_key)
            yield session

    # User doesn't have a session associated with it
    except AuthUserSession.DoesNotExist:
        # Create session for user
        session = SessionStore()
        session.create()
        session_object = Session.objects.get(pk=session.session_key)

        # Save AuthUserSession for user
        AuthUserSession(auth_user_id=user_id, session_id=session_object).save()
        yield session

    # After context manager goes out of scope, save session data
    finally:
        session.save()
```

**Comment by gannonprudhomme** [comment]
Also can you merge master into this? You have some conflicts rn

**Comment by ThatJuanGuy** [comment]
Do not merge this branch before logout button

**Comment by ThatJuanGuy** [comment]
I merged logout into this branch so that yall can see/test it in the state it will be once it gets merged into master

**Review by ryan-conn — CHANGES_REQUESTED** [review]
A couple of suggestions to make the code a bit cleaner, the main thing I really want in this is saving the anonymous session to the user's session data when you log in to prevent accidental data loss

File: autoscheduler/user_sessions/utils/retrieve_data_session.py
Comment (gannonprudhomme) [reply autoscheduler/user_sessions/utils/retrieve_data_session.py:32]: What about the exact same situation, but you've logged in previously and have session data for that term? Should we _always_ overwrite the logged in data with the anonymous session data?

Comment (ryan-conn) [reply autoscheduler/user_sessions/utils/retrieve_data_session.py:32]: I don't think so, seems like that would do far more harm than good and I can't imagine a case where users would want to do this unless they just forgot to log in

File: autoscheduler/user_sessions/utils/retrieve_data_session.py
Comment (ryan-conn) [inline autoscheduler/user_sessions/utils/retrieve_data_session.py:16]:
```suggestion
    user_id = request.user.id
```
One more simplifaction and this should be ready to merge, thought I pointed it out in the last round of review but I forgot to oops

**Review by ryan-conn — CHANGES_REQUESTED** [review]
Tested that the new functionality is working, only thing is a comment I forgot to make on the last round

**Review by ryan-conn — APPROVED** [review]

**Pull request merged**
